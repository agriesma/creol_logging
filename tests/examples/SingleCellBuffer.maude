load creol-interpreter .
mod PROGRAM is
  protecting CREOL-SIMULATOR .
  op classes : -> Configuration [ctor] .
  eq classes =
    < "Buffer" : Class | Inh: noInh, Param: noVid, Att: "cell" |-> null,
      "ctr" |-> null, "next" |-> null,
      Mtds: < "init" : Method | Param: noVid, Att: "_" |-> null,
              Code: assign( "ctr" ; int(0) ) ; return ( emp ) >,
            < "run" : Method | Param: noVid, Att: "_" |-> null, Code: skip ;
              return ( emp ) >,
            < "put" : Method | Param: "x", Att: "x" |-> null, "label:0" |->
              null, "_" |-> null, Code: if "=" ( "ctr" :: int(0) ) th
              assign( "cell" ; "x" ) el if "=" ( "next" :: null ) th
              new( "next" ; "Buffer" ; emp ) el skip fi ; call( "label:0" ;
              "next" ; "put" ; "x" ) ; get( "label:0" ; noVid ) ;
              free( "label:0" ) fi ; assign( "ctr" ; "+" ( "ctr" ::
              int(1) ) ) ; return ( emp ) >,
            < "get" : Method | Param: noVid, Att: "x" |-> null, "label:0" |->
              null, "_" |-> null, Code: await ">" ( "ctr" :: int(0) ) ; if
              "/=" ( "cell" :: null ) th assign( "x" ; "cell" ) el
              call( "label:0" ; "next" ; "get" ; emp ) ;
              get( "label:0" ; "x" ) ; free( "label:0" ) fi ;
              assign( "ctr" ; "-" ( "ctr" :: int(1) ) ) ;
              assign( "cell" ; null ) ; return ( "x" ) >,
      Ocnt: 0 > .
endm

