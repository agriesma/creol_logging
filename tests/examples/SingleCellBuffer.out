==========================================
reduce in PROGRAM : classes .
result Class: 
< "Buffer" : Class | Inh: noInh, Param: noVid, Att: "cell" |-> null, "ctr" |->
    null, "next" |-> null, Mtds:
  < "get" : Method | Param: noVid, Att: "_" |-> null, "label:0" |-> null, "x"
    |-> null, Code: await ">"("ctr" :: int(0)) ; if "/="("cell" :: null) th
    assign("x" ; "cell") el call("label:0" ; "next" ; "get" ; emp) ; get(
    "label:0" ; "x") ; free("label:0") fi ; assign("ctr" ; "-"("ctr" :: int(
    1))) ; assign("cell" ; null) ; return("x") >, < "init" : Method | Param:
    noVid, Att: "_" |-> null, Code: assign("ctr" ; int(0)) ; return(emp) >, <
    "put" : Method | Param: "x", Att: "_" |-> null, "label:0" |-> null, "x" |->
    null, Code: if "="("ctr" :: int(0)) th assign("cell" ; "x") el if "="(
    "next" :: null) th new("next" ; "Buffer" ; emp) el skip fi ; call("label:0"
    ; "next" ; "put" ; "x") ; get("label:0" ; noVid) ; free("label:0") fi ;
    assign("ctr" ; "+"("ctr" :: int(1))) ; return(emp) >, < "run" : Method |
    Param: noVid, Att: "_" |-> null, Code: skip ; return(emp) >,
  Ocnt: 0 >

Bye.
